src/fd.cpp

```diff
@@ -0,0 +1,264 @@
+
+}
+
+// CONSIDER: if necessary to add method for HANDLES on Windows. Would be different code to SOCKETs.
+// TODO: implement re-usable background thread.
+static void *select_thread(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  int result;
+  int *values;
+
+#ifndef POLLIN // fall back to select() for R on older Windows
+
+  fd_set readfds, writefds, exceptfds;
+  FD_ZERO(&readfds);
+  FD_ZERO(&writefds);
+  FD_ZERO(&exceptfds);
+  int max_fd = *std::max_element(args->fds->begin(), args->fds->end());
+  for (int i = 0; i < args->rfds; i++) {
+    FD_SET((*args->fds)[i], &readfds);
+  }
+  for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+    FD_SET((*args->fds)[i], &writefds);
+  }
+  for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+    FD_SET((*args->fds)[i], &exceptfds);
+  }
+
+  struct timeval tv;
+  int use_timeout = args->timeout != R_PosInf;
+  if (use_timeout) {
+    tv.tv_sec = args->timeout < 0 ? 1 : (int) args->timeout;
+    tv.tv_usec = args->timeout < 0 ? 0 : ((int) (args->timeout * 1000)) % 1000 * 1000;
+  }
+
+  result = select(max_fd + 1, &readfds, &writefds, &exceptfds, use_timeout ? &tv : NULL);
+
+  values = (int *) DATAPTR_RO(CADR(args->callback));
+  if (result > 0) {
+    for (int i = 0; i < args->rfds; i++) {
+      values[i] = FD_ISSET((*args->fds)[i], &readfds);
+    }
+    for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+      FD_ISSET((*args->fds)[i], &writefds);
+    }
+    for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+      FD_ISSET((*args->fds)[i], &exceptfds);
+    }
+  } else {
+    for (int i = 0; i < args->num_fds; i++) {
+      values[i] = result == 0 ? 0 : R_NaInt;
+    }
+  }
+
+#else
+
+  int timeout = args->timeout == R_PosInf ? -1 : args->timeout < 0 ? 1000 : (int) (args->timeout * 1000);
+
+  std::vector<struct pollfd> pollfds;
+  pollfds.reserve(args->num_fds);
+  struct pollfd pfd;
+  for (int i = 0; i < args->rfds; i++) {
+    pfd.fd = (*args->fds)[i];
+    pfd.events = POLLIN;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
+    pfd.fd = (*args->fds)[i];
+    pfd.events = POLLOUT;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+  for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
+    pfd.fd = (*args->fds)[i];
+    pfd.events = 0;
+    pfd.revents = 0;
+    pollfds.push_back(pfd);
+  }
+#ifdef _WIN32
+  result = WSAPoll(pollfds.data(), args->num_fds, timeout);
+#else
+  result = poll(pollfds.data(), args->num_fds, timeout);
+#endif
+
+  values = (int *) DATAPTR_RO(CADR(args->callback));
+  if (result > 0) {
+    for (int i = 0; i < args->num_fds; i++) {
+      values[i] = pollfds[i].revents & (POLLIN | POLLOUT) ? 1 : pollfds[i].revents & (POLLNVAL | POLLHUP | POLLERR) ? R_NaInt : 0;
+    }
+  } else {
```

Comment:
Are you consciously not distinguishing between timeout (`result == 0`) and error (`result < 0`)?

Reply:
I do just a couple of lines lower: `values[i] = result == 0 ? 0 : R_NaInt;`
Can move this conditional up to the top level if you think it's clearer.

Reply:
I'm amended in [3b982bd](https://github.com/r-lib/later/pull/190/commits/3b982bd5664314ea1c5edc51b45fdb00578ce6ff), which also handles the timeout case marginally more efficiently...

Reply:
Oh, I see. I was looking for `errno`/`WSAGetLastError` handling. I just want to make sure sensible things happen in foreseeable error scenarios, like the network being interrupted for whatever reason, or especially, the main R thread deciding to close the curl/httr2 request.

Reply:
I now saw [your earlier comment](https://github.com/r-lib/later/pull/190#discussion_r1813340780) that you tested closing before `later_fd`. If network interruption also does something sensible on both Posix and Windows, you can resolve this conversation.

Reply:
Yes, I try really hard to ensure the background thread always calls the later callback so errors are returned as `NA` values (it's not that useful to error on a background thread!).

I've tweaked the return values a bit in [1ac44cc](https://github.com/shikokuchuo/later/commit/1ac44cc286f6ce5cf1c129c1dcd9dcabdfb54e4c): This ensures that the `exceptfds` return `TRUE` in case they are triggered rather than a `NA`, as that is the expectation.

I think this conv is now resolved but I don't see a way to do that myself... do I need to be a member of r-lib? :p